deps/rabbit/src/rabbit_binding.erl

```diff
@@ -100,196 +90,107 @@ recover(XNames, QNames) ->
                 destination = #resource{kind = queue, name = QName} = Queue,
                 key = QName,
                 args = []}) ->
     rabbit_amqqueue:exists(Queue);
-exists(Binding) ->
-    binding_action(
-      Binding, fun (_Src, _Dst, B) ->
-                       rabbit_misc:const(mnesia:read({rabbit_route, B}) /= [])
-               end, fun not_found_or_absent_errs/1).
+exists(Binding0) ->
+    Binding = sort_args(Binding0),
+    rabbit_db_binding:exists(Binding).
 
 -spec add(rabbit_types:binding(), rabbit_types:username()) -> bind_res().
 
 add(Binding, ActingUser) -> add(Binding, fun (_Src, _Dst) -> ok end, ActingUser).
 
 -spec add(rabbit_types:binding(), inner_fun(), rabbit_types:username()) -> bind_res().
 
-add(Binding, InnerFun, ActingUser) ->
-    binding_action(
-      Binding,
-      fun (Src, Dst, B) ->
-              case rabbit_exchange:validate_binding(Src, B) of
-                  ok ->
-                      lock_resource(Src, read),
-                      lock_resource(Dst, read),
-                      %% this argument is used to check queue exclusivity;
-                      %% in general, we want to fail on that in preference to
-                      %% anything else
-                      case InnerFun(Src, Dst) of
-                          ok ->
-                              case mnesia:read({rabbit_route, B}) of
-                                  []  -> add(Src, Dst, B, ActingUser);
-                                  [_] -> fun () -> ok end
-                              end;
-                          {error, _} = Err ->
-                              rabbit_misc:const(Err)
-                      end;
-                  {error, _} = Err ->
-                      rabbit_misc:const(Err)
-              end
-      end, fun not_found_or_absent_errs/1).
-
-add(Src, Dst, B, ActingUser) ->
-    ok = sync_route(#route{binding = B}, durable(Src), durable(Dst),
-                    should_index_table(Src), fun mnesia:write/3),
-    x_callback(transaction, Src, add_binding, B),
-    Serial = rabbit_exchange:serial(Src),
-    fun () ->
-        x_callback(Serial, Src, add_binding, B),
-        ok = rabbit_event:notify(
-            binding_created,
-            info(B) ++ [{user_who_performed_action, ActingUser}])
+add(Binding0, InnerFun, ActingUser) ->
+    Binding = sort_args(Binding0),
+    case
+        rabbit_db_binding:create(Binding, binding_checks(Binding, InnerFun))
+    of
+        ok ->
+            ok = rabbit_event:notify(
+                   binding_created,
+                   info(Binding) ++ [{user_who_performed_action, ActingUser}]);
+        Err ->
+            Err
     end.
 
--spec remove(rabbit_types:binding()) -> bind_res().
-remove(Binding) -> remove(Binding, fun (_Src, _Dst) -> ok end, ?INTERNAL_USER).
-
--spec remove(rabbit_types:binding(), rabbit_types:username()) -> bind_res().
-remove(Binding, ActingUser) -> remove(Binding, fun (_Src, _Dst) -> ok end, ActingUser).
```

Comment:
Looks like this is used in `rabbit_mqtt_processor:unbind/3` (might be new from the merge/rebase on #5895 I think?)

Reply:
Yes, it looks like is used by the processor now. Restored.